### ISSUE TITLE
Functions do not match callback protocol with `ParamSpec`

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#2354]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from typing import Any, Generic, Protocol
from typing_extensions import ParamSpec

P = ParamSpec('P')


class CommandHandler(Protocol[P]):
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> dict[str, Any]:
        ...


class Command(Generic[P]):
    __slots__ = ('handler',)

    def __init__(self, handler: CommandHandler[P]) -> None:
        self.handler: CommandHandler[P] = handler


def handler(arg1: int, arg2: str) -> dict[str, Any]:
    return {}


# Error: Argument of type "(arg1: int, arg2: str) -> dict[str, Any]" cannot be assigned
# to parameter "handler" of type "CommandHandler[P@Command]" in function "__init__"
#   Type "(arg1: int, arg2: str) -> dict[str, Any]" cannot be assigned to type
#   "Callable[Concatenate[P.args, P.kwargs, P@Command], dict[str, Any]]"
#       Parameter "*args" has no corresponding parameter
command = Command(handler)


command  # type is Command[P@Command]
command.handler  # type is CommandHandler[P@Command]
```

An error is produced as detailed in the comment above and the types of `command` and `command.handler` are still generic.

**Expected behavior**
No error should be produced, the type of `command` should be `Command[(arg1: int, arg2: int)]`, and the type of `command.handler` should be `(arg1: int, arg2: str) -> dict[str, Any]`.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.172

**Additional context**
If `CommandHandler` is changed to from a callback protocol to `Callable[P, dict[str, Any]]`, the code produces the expected behavior.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.173, which I just published. It will also be included in the next release of pylance.
